### DISCUSSION TITLE
Http responsibility in the domain?

### DISCUSSION THREAD

**Discussion by marc-bouvier** [juanm4/hexagonal-architecture-frontend#3] [General]
Hi, I am a bit surprised to see Http concerns in the domain part of the sample. It seems to me such technical concept (communication protocol), might be located in the infrastructure part. 

We could decide to switch to graphql or websocket and it should not impact the domain part in my opinion.

What are your thoughts about this?

Kindly,

Marc

**Comment by juanm4** [comment]
Hi again Marc, after a quick review...I agree with you. 

Maybe in a real case the best option is forget about the `Http` interface and only implement the interface `ProductRepository` without passing any client. Then implement the two methods `getProducts` and `getProductsById` using the client we need. 

Sometimes we tend to make things more complicated that we need because we are thinking in the future. In this case **I wanted to show how to use dependency injection and composition** using a http client as parameter. 

But in the practice, I would forget about `Http` and I would implement `ProductRepository` as a service. I mean, if I'm going to use http protocol I will implement a `ProductRepository` that use http...If I'm going to use graphql I would implement a `ProductRepository` that use graphql...and so on. Then you can choose at any time what implementation of `ProductRepository` you need and change one by another very quickly.

What do you think?
All the best!
Juanma

**Reply by marc-bouvier** [reply]
After 2 years, I tend to have an approach a bit more pragmatic. And I'll happily keep http client concern in the domain if it reduces cognitive load and does not prevent testing.

By injecting the HTTP client it would not be that hard to extract an interface and abstract http infrastructure details if needed.

Also, with technologies such as Mock Service Workers, we can in some cases test such components with a less fragile approach that traditional mocks.
